File: apps/android-app/src/main/java/com/fire/app/richtext/FireRenderPresentation.kt
package com.fire.app.richtext

import uniffi.fire_uniffi_types.RenderDocumentHandle
import uniffi.fire_uniffi_types.RenderOneboxCardState
import uniffi.fire_uniffi_types.RenderRichNodeState
import uniffi.fire_uniffi_types.RenderUiSegmentState

/// Host mapping for Rust `RenderDocumentHandle`.
///
/// Production cells consume the precomputed UI plan. They do not parse
/// cooked HTML or rebuild a block tree.
object FireRenderPresentation {
    fun content(handle: RenderDocumentHandle): FireRichTextContent {
        return FireRichTextContent(
            nodes = richNodes(handle),
            plainText = handle.plainText(),
            imageAttachments = images(handle),
        )
    }

    fun blocks(handle: RenderDocumentHandle): List<FireRichTextBlock> {
        return (0u until handle.segmentCount()).mapNotNull { index ->
            when (val segment = handle.segment(index)) {
                is RenderUiSegmentState.Rich -> {
                    val nodes = segment.nodes.map(::mapNode)
                    if (nodes.isEmpty()) null else FireRichTextBlock.Text(nodes)
                }
                is RenderUiSegmentState.Image -> {
                    val url = segment.image.url.trim()
                    if (url.isEmpty()) {
                        null
                    } else {
                        FireRichTextBlock.Image(
                            FireCookedImage(
                                url = url,
                                altText = segment.image.altText,
                                width = segment.image.width?.toFloat(),
                                height = segment.image.height?.toFloat(),
                            ),
                        )
                    }
                }
                is RenderUiSegmentState.Onebox -> {
                    FireRichTextBlock.Text(listOf(mapOnebox(segment.card)))
                }
                null -> null
            }
        }
    }

    fun images(handle: RenderDocumentHandle): List<FireCookedImage> {
        return handle.imageAttachments().map { image ->
            FireCookedImage(
                url = image.url,
                altText = image.altText,
                width = image.width?.toFloat(),
                height = image.height?.toFloat(),
            )
        }
    }

    fun richNodes(handle: RenderDocumentHandle): List<FireRichTextNode> {
        return (0u until handle.segmentCount()).flatMap { index ->
            when (val segment = handle.segment(index)) {
                is RenderUiSegmentState.Rich -> segment.nodes.map(::mapNode)
                else -> emptyList()
            }
        }
    }

    fun mapNode(node: RenderRichNodeState): FireRichTextNode {
        return when (node) {
            is RenderRichNodeState.Text -> FireRichTextNode.Text(node.content)
            is RenderRichNodeState.Bold -> FireRichTextNode.Bold(node.children.map(::mapNode))
            is RenderRichNodeState.Italic -> FireRichTextNode.Italic(node.children.map(::mapNode))
            is RenderRichNodeState.Strikethrough -> FireRichTextNode.Strikethrough(node.children.map(::mapNode))
            is RenderRichNodeState.Code -> FireRichTextNode.Code(node.code)
            is RenderRichNodeState.CodeBlock -> FireRichTextNode.CodeBlock(node.language, node.code)
            is RenderRichNodeState.Link -> FireRichTextNode.Link(node.url, node.children.map(::mapNode))
            is RenderRichNodeState.Mention -> FireRichTextNode.Mention(node.username)
            is RenderRichNodeState.MentionGroup -> FireRichTextNode.MentionGroup(node.name, node.url)
            is RenderRichNodeState.Hashtag -> FireRichTextNode.Hashtag(node.text, node.url, node.kind)
            is RenderRichNodeState.Emoji -> FireRichTextNode.Emoji(node.url, node.fallbackText, node.onlyEmoji)
            is RenderRichNodeState.Heading -> FireRichTextNode.Heading(node.level.toInt(), node.children.map(::mapNode))
            is RenderRichNodeState.Blockquote -> FireRichTextNode.Blockquote(node.children.map(::mapNode))
            is RenderRichNodeState.Quote -> FireRichTextNode.Quote(
                author = node.author,
                postNumber = node.postNumber,
                topicId = node.topicId,
                children = node.children.map(::mapNode),
            )
            is RenderRichNodeState.ListNode -> FireRichTextNode.ListNode(
                ordered = node.ordered,
                items = node.items.map { item -> item.map(::mapNode) },
            )
            is RenderRichNodeState.ListItem -> FireRichTextNode.ListItem(node.children.map(::mapNode))
            is RenderRichNodeState.Spoiler -> FireRichTextNode.Spoiler(node.children.map(::mapNode))
            is RenderRichNodeState.Details -> FireRichTextNode.Details(
                summary = node.summary.map(::mapNode),
                children = node.children.map(::mapNode),
            )
            is RenderRichNodeState.Table -> FireRichTextNode.Table(node.text)
            is RenderRichNodeState.Video -> FireRichTextNode.Video(node.url, node.title)
            is RenderRichNodeState.Divider -> FireRichTextNode.Divider
            is RenderRichNodeState.LineBreak -> FireRichTextNode.LineBreak
            is RenderRichNodeState.Paragraph -> FireRichTextNode.Paragraph(node.children.map(::mapNode))
            is RenderRichNodeState.Image -> FireRichTextNode.Image(
                src = node.url,
                alt = node.alt,
                width = node.width?.toFloat(),
                height = node.height?.toFloat(),
            )
            is RenderRichNodeState.Onebox -> mapOnebox(node.card)
        }
    }

    private fun mapOnebox(card: RenderOneboxCardState): FireRichTextNode.Onebox {
        return FireRichTextNode.Onebox(
            url = card.url,
            title = card.title,
            description = card.description,
            sourceName = card.sourceName,
            iconUrl = card.iconUrl,
            thumbnailUrl = card.thumbnailUrl,
            thumbnailWidth = card.thumbnailWidth?.toFloat(),
            thumbnailHeight = card.thumbnailHeight?.toFloat(),
        )
    }
}

File: apps/ios-app/App/Views/Shared/FireRenderPresentation.swift
import Foundation
import UIKit

/// Host mapping for Rust `RenderDocumentHandle` / UI plan.
///
/// Platforms only turn precomputed segments into native nodes. They do not
/// parse cooked HTML or walk a block tree.
enum FireRenderPresentation {
    static func checksum(_ handle: RenderDocumentHandle) -> UInt64 {
        handle.checksum()
    }

    static func images(from handle: RenderDocumentHandle) -> [FireCookedImage] {
        handle.imageAttachments().compactMap(cookedImage(from:))
    }

    static func richNodes(from handle: RenderDocumentHandle) -> [FireRichTextNode] {
        segments(from: handle).flatMap { segment -> [FireRichTextNode] in
            switch segment {
            case let .rich(nodes):
                return nodes
            case .image, .onebox:
                return []
            }
        }
    }

    static func segments(from handle: RenderDocumentHandle) -> [MappedUiSegment] {
        let count = handle.segmentCount()
        return (UInt32(0)..<count).compactMap { index in
            handle.segment(index: index).map(mapSegment)
        }
    }

    static func cookedImage(from image: RenderImageAttachmentState) -> FireCookedImage? {
        guard let url = URL(string: image.url) else {
            return nil
        }
        return FireCookedImage(
            url: url,
            altText: image.altText,
            width: image.width.map(CGFloat.init),
            height: image.height.map(CGFloat.init)
        )
    }

    enum MappedUiSegment {
        case rich([FireRichTextNode])
        case image(FireCookedImage)
        case onebox(FireTopicOneboxCard)
    }

    static func mapSegment(_ segment: RenderUiSegmentState) -> MappedUiSegment {
        switch segment {
        case let .rich(nodes):
            return .rich(nodes.map(mapNode))
        case let .image(image):
            let cooked = cookedImage(from: image) ?? FireCookedImage(
                url: URL(string: image.url) ?? URL(string: "about:blank")!,
                altText: image.altText,
                width: image.width.map(CGFloat.init),
                height: image.height.map(CGFloat.init)
            )
            return .image(cooked)
        case let .onebox(card):
            return .onebox(oneboxCard(from: card))
        }
    }

    static func oneboxCard(from card: RenderOneboxCardState) -> FireTopicOneboxCard {
        FireTopicOneboxCard(
            url: card.url,
            title: card.title,
            description: card.description,
            sourceName: card.sourceName,
            iconURL: card.iconUrl.flatMap(URL.init(string:)),
            thumbnailURL: card.thumbnailUrl.flatMap(URL.init(string:)),
            thumbnailWidth: card.thumbnailWidth.map(CGFloat.init),
            thumbnailHeight: card.thumbnailHeight.map(CGFloat.init)
        )
    }

    static func mapNode(_ node: RenderRichNodeState) -> FireRichTextNode {
        switch node {
        case let .text(content):
            return .text(content)
        case let .bold(children):
            return .bold(children.map(mapNode))
        case let .italic(children):
            return .italic(children.map(mapNode))
        case let .strikethrough(children):
            return .strikethrough(children.map(mapNode))
        case let .code(code):
            return .code(code)
        case let .codeBlock(language, code):
            return .codeBlock(language: language, code: code)
        case let .link(url, children):
            return .link(url: url, children: children.map(mapNode))
        case let .mention(username):
            return .mention(username: username)
        case let .mentionGroup(name, url):
            return .mentionGroup(name: name, url: url)
        case let .hashtag(text, url, kind):
            return .hashtag(text: text, url: url, kind: kind)
        case let .emoji(url, fallbackText, onlyEmoji):
            return .emoji(url: url, fallbackText: fallbackText, onlyEmoji: onlyEmoji)
        case let .heading(level, children):
            return .heading(level: Int(level), children: children.map(mapNode))
        case let .blockquote(children):
            return .blockquote(children.map(mapNode))
        case let .quote(author, postNumber, topicId, children):
            return .quote(
                author: author,
                postNumber: postNumber,
                topicId: topicId,
                children: children.map(mapNode)
            )
        case let .listNode(ordered, items):
            return .list(ordered: ordered, items: items.map { $0.map(mapNode) })
        case let .listItem(children):
            return .listItem(children.map(mapNode))
        case let .spoiler(children):
            return .spoiler(children.map(mapNode))
        case let .details(summary, children):
            return .details(summary: summary.map(mapNode), children: children.map(mapNode))
        case let .table(text):
            return .table(text)
        case let .video(url, title):
            return .video(url: url, title: title)
        case .divider:
            return .divider
        case .lineBreak:
            return .lineBreak
        case let .paragraph(children):
            return .paragraph(children.map(mapNode))
        case let .image(url, alt, width, height):
            return .image(
                src: url,
                alt: alt,
                width: width.map(CGFloat.init),
                height: height.map(CGFloat.init)
            )
        case let .onebox(card):
            return .onebox(
                url: card.url,
                title: card.title,
                description: card.description,
                sourceName: card.sourceName,
                iconURL: card.iconUrl,
                thumbnailURL: card.thumbnailUrl,
                thumbnailWidth: card.thumbnailWidth,
                thumbnailHeight: card.thumbnailHeight
            )
        }
    }
}
